### PR TITLE
[RV64_DYNAREC] Support HSUBPS and MOVDQ2Q instruction

### PR DESCRIPTION
**Pull request by zengdage**

### CODE REVIEW

**Review by ptitSeb — APPROVED** [review]

**Comment by ksco** [comment]
I fixed a regression yesterday: #4131, so please provide additional details: What program uses these opcodes? How are they tested?

**Comment by zengdage** [comment]
> I fixed a regression yesterday: #4131, so please provide additional details: What program uses these opcodes? How are they tested?

Sorry again for this issue in #4131.

HSUBPS and MOVDQ2Q are used in wemeet applications.
```
./lib/libQt5WebEngineCore.s:12381364: 360ff43:  f2 0f d6 c4             movdq2q %xmm4,%mm0
./lib/libQt5WebEngineCore.s:12381366: 360ff4a:  f2 0f d6 cc             movdq2q %xmm4,%mm1
./lib/libQt5WebEngineCore.s:12381367: 360ff4e:  f2 0f d6 d5             movdq2q %xmm5,%mm2
./lib/libQt5WebEngineCore.s:12381369: 360ff55:  f2 0f d6 dd             movdq2q %xmm5,%mm3
./lib/libQt5WebEngineCore.s:12381370: 360ff59:  f2 0f d6 ee             movdq2q %xmm6,%mm5
./lib/libQt5WebEngineCore.s:12381372: 360ff60:  f2 0f d6 f6             movdq2q %xmm6,%mm6
```
```
./lib/libQt5WebEngineCore.s:12346399: 35ef90f:  f2 0f 7d d8             hsubps %xmm0,%xmm3
```

My Tests.
```
// Test F2 0F: HSUBPS Gx, Ex and MOVDQ2Q Gm, Ex
// Build: gcc -msse3 -O2 -o test_f20f_hsubps_movdq2q test_f20f_hsubps_movdq2q.c
// Run:   ./test_f20f_hsubps_movdq2q
// Under box64: box64 ./test_f20f_hsubps_movdq2q
#include <stdio.h>
#include <stdint.h>
#include <string.h>

static int tests_run = 0;
static int tests_failed = 0;

#define CHECK(name, got, expected)                                          \
    do {                                                                    \
        tests_run++;                                                        \
        uint64_t g_ = (uint64_t)(got);                                      \
        uint64_t e_ = (uint64_t)(expected);                                 \
        if (g_ != e_) {                                                     \
            printf("FAIL %-45s got 0x%016llx expected 0x%016llx\n",        \
                   name, (unsigned long long)g_, (unsigned long long)e_);   \
            tests_failed++;                                                 \
        }                                                                   \
    } while (0)

// Compare two floats by bit pattern
#define CHECK_F32(name, got, expected)                                      \
    do {                                                                    \
        tests_run++;                                                        \
        uint32_t g_, e_;                                                    \
        memcpy(&g_, &(got), 4);                                             \
        memcpy(&e_, &(expected), 4);                                        \
        if (g_ != e_) {                                                     \
            printf("FAIL %-45s got %.6f (0x%08x) expected %.6f (0x%08x)\n",\
                   name, (double)(got), g_, (double)(expected), e_);        \
            tests_failed++;                                                 \
        }                                                                   \
    } while (0)

typedef float  v4f32 __attribute__((vector_size(16), aligned(16)));
typedef uint32_t v4u32 __attribute__((vector_size(16), aligned(16)));

// ===================== HSUBPS =====================
// HSUBPS Gx, Ex:
//   GX->f[0] = GX->f[0] - GX->f[1]
//   GX->f[1] = GX->f[2] - GX->f[3]
//   GX->f[2] = EX->f[0] - EX->f[1]
//   GX->f[3] = EX->f[2] - EX->f[3]
static void test_hsubps(void)
{
    __attribute__((aligned(16))) float gx[4], ex[4], res[4];

    // Basic: gx = {1,2,3,4}, ex = {5,6,7,8}
    // result = {1-2, 3-4, 5-6, 7-8} = {-1, -1, -1, -1}
    gx[0]=1.0f; gx[1]=2.0f; gx[2]=3.0f; gx[3]=4.0f;
    ex[0]=5.0f; ex[1]=6.0f; ex[2]=7.0f; ex[3]=8.0f;
    __asm__ volatile(
        "movaps %1, %%xmm0\n\t"
        "hsubps %2, %%xmm0\n\t"
        "movaps %%xmm0, %0\n\t"
        : "=m"(*res) : "m"(*gx), "m"(*ex) : "xmm0"
    );
    float m1 = -1.0f;
    CHECK_F32("HSUBPS basic res[0]", res[0], m1);
    CHECK_F32("HSUBPS basic res[1]", res[1], m1);
    CHECK_F32("HSUBPS basic res[2]", res[2], m1);
    CHECK_F32("HSUBPS basic res[3]", res[3], m1);

    // Asymmetric: gx = {10,3,8,2}, ex = {100,40,9,1}
    // result = {10-3, 8-2, 100-40, 9-1} = {7, 6, 60, 8}
    gx[0]=10.0f; gx[1]=3.0f; gx[2]=8.0f; gx[3]=2.0f;
    ex[0]=100.0f; ex[1]=40.0f; ex[2]=9.0f; ex[3]=1.0f;
    __asm__ volatile(
        "movaps %1, %%xmm1\n\t"
        "hsubps %2, %%xmm1\n\t"
        "movaps %%xmm1, %0\n\t"
        : "=m"(*res) : "m"(*gx), "m"(*ex) : "xmm1"
    );
    float e7=7.0f, e6=6.0f, e60=60.0f, e8=8.0f;
    CHECK_F32("HSUBPS asym res[0]", res[0], e7);
    CHECK_F32("HSUBPS asym res[1]", res[1], e6);
    CHECK_F32("HSUBPS asym res[2]", res[2], e60);
    CHECK_F32("HSUBPS asym res[3]", res[3], e8);

    // Self-operand: gx == ex = {1,2,3,4}
    // result = {1-2, 3-4, 1-2, 3-4} = {-1, -1, -1, -1}
    gx[0]=1.0f; gx[1]=2.0f; gx[2]=3.0f; gx[3]=4.0f;
    __asm__ volatile(
        "movaps %1, %%xmm2\n\t"
        "hsubps %%xmm2, %%xmm2\n\t"
        "movaps %%xmm2, %0\n\t"
        : "=m"(*res) : "m"(*gx) : "xmm2"
    );
    CHECK_F32("HSUBPS self res[0]", res[0], m1);
    CHECK_F32("HSUBPS self res[1]", res[1], m1);
    CHECK_F32("HSUBPS self res[2]", res[2], m1);
    CHECK_F32("HSUBPS self res[3]", res[3], m1);

    // Zero: gx = {0,0,0,0}, ex = {0,0,0,0} → all zeros
    gx[0]=0.0f; gx[1]=0.0f; gx[2]=0.0f; gx[3]=0.0f;
    ex[0]=0.0f; ex[1]=0.0f; ex[2]=0.0f; ex[3]=0.0f;
    __asm__ volatile(
        "movaps %1, %%xmm3\n\t"
        "hsubps %2, %%xmm3\n\t"
        "movaps %%xmm3, %0\n\t"
        : "=m"(*res) : "m"(*gx), "m"(*ex) : "xmm3"
    );
    float z = 0.0f;
    CHECK_F32("HSUBPS zero res[0]", res[0], z);
    CHECK_F32("HSUBPS zero res[1]", res[1], z);
    CHECK_F32("HSUBPS zero res[2]", res[2], z);
    CHECK_F32("HSUBPS zero res[3]", res[3], z);

    // Negative inputs: gx = {-1,-2,-3,-4}, ex = {-5,-6,-7,-8}
    // result = {-1-(-2), -3-(-4), -5-(-6), -7-(-8)} = {1, 1, 1, 1}
    gx[0]=-1.0f; gx[1]=-2.0f; gx[2]=-3.0f; gx[3]=-4.0f;
    ex[0]=-5.0f; ex[1]=-6.0f; ex[2]=-7.0f; ex[3]=-8.0f;
    __asm__ volatile(
        "movaps %1, %%xmm4\n\t"
        "hsubps %2, %%xmm4\n\t"
        "movaps %%xmm4, %0\n\t"
        : "=m"(*res) : "m"(*gx), "m"(*ex) : "xmm4"
    );
    float p1 = 1.0f;
    CHECK_F32("HSUBPS neg res[0]", res[0], p1);
    CHECK_F32("HSUBPS neg res[1]", res[1], p1);
    CHECK_F32("HSUBPS neg res[2]", res[2], p1);
    CHECK_F32("HSUBPS neg res[3]", res[3], p1);
}

// ===================== MOVDQ2Q =====================
// MOVDQ2Q Gm, Ex: copy low 64 bits of XMM register to MMX register
// We verify by storing the MMX register to memory and checking the value.
static void test_movdq2q(void)
{
    __attribute__((aligned(16))) uint64_t xmm_data[2];
    __attribute__((aligned(8)))  uint64_t mmx_result;

    // Pattern 1: low64 = 0xDEADBEEFCAFEBABE, high64 = 0x1234567890ABCDEF
    xmm_data[0] = UINT64_C(0xDEADBEEFCAFEBABE);
    xmm_data[1] = UINT64_C(0x1234567890ABCDEF);
    mmx_result = 0;
    __asm__ volatile(
        "movdqa %1, %%xmm5\n\t"
        "movdq2q %%xmm5, %%mm0\n\t"
        "movq %%mm0, %0\n\t"
        "emms\n\t"
        : "=m"(mmx_result)
        : "m"(*xmm_data)
        : "xmm5", "mm0"
    );
    CHECK("MOVDQ2Q low64 pattern1", mmx_result, UINT64_C(0xDEADBEEFCAFEBABE));

    // Pattern 2: low64 = 0x0000000000000001, high64 = 0xFFFFFFFFFFFFFFFF
    xmm_data[0] = UINT64_C(0x0000000000000001);
    xmm_data[1] = UINT64_C(0xFFFFFFFFFFFFFFFF);
    mmx_result = 0;
    __asm__ volatile(
        "movdqa %1, %%xmm6\n\t"
        "movdq2q %%xmm6, %%mm1\n\t"
        "movq %%mm1, %0\n\t"
        "emms\n\t"
        : "=m"(mmx_result)
        : "m"(*xmm_data)
        : "xmm6", "mm1"
    );
    CHECK("MOVDQ2Q low64 pattern2", mmx_result, UINT64_C(0x0000000000000001));

    // Pattern 3: low64 = 0, high64 = 0xDEADBEEFCAFEBABE
    // MMX should get 0, not the high 64 bits
    xmm_data[0] = UINT64_C(0x0000000000000000);
    xmm_data[1] = UINT64_C(0xDEADBEEFCAFEBABE);
    mmx_result = UINT64_C(0xFFFFFFFFFFFFFFFF);
    __asm__ volatile(
        "movdqa %1, %%xmm7\n\t"
        "movdq2q %%xmm7, %%mm2\n\t"
        "movq %%mm2, %0\n\t"
        "emms\n\t"
        : "=m"(mmx_result)
        : "m"(*xmm_data)
        : "xmm7", "mm2"
    );
    CHECK("MOVDQ2Q low64 zero (not high)", mmx_result, UINT64_C(0x0000000000000000));

    // Pattern 4: all-ones low64
    xmm_data[0] = UINT64_C(0xFFFFFFFFFFFFFFFF);
    xmm_data[1] = UINT64_C(0x0000000000000000);
    mmx_result = 0;
    __asm__ volatile(
        "movdqa %1, %%xmm0\n\t"
        "movdq2q %%xmm0, %%mm3\n\t"
        "movq %%mm3, %0\n\t"
        "emms\n\t"
        : "=m"(mmx_result)
        : "m"(*xmm_data)
        : "xmm0", "mm3"
    );
    CHECK("MOVDQ2Q low64 all-ones", mmx_result, UINT64_C(0xFFFFFFFFFFFFFFFF));
}

int main(void)
{
    printf("=== F2 0F: HSUBPS / MOVDQ2Q test (inline asm) ===\n\n");

    printf("[HSUBPS Gx, Ex]\n");   test_hsubps();
    printf("[MOVDQ2Q Gm, Ex]\n");  test_movdq2q();

    printf("\n%d tests, %d failed\n", tests_run, tests_failed);
    return tests_failed ? 1 : 0;
}
```

**Review by ksco — APPROVED** [review]

**Comment by zengdage** [comment]
Some problems remain for `Inf` and `NaN` edge cases for `HSUBPS`. I will address them first.

**Comment by zengdage** [comment]
> Some problems remain for `Inf` and `NaN` edge cases for `HSUBPS`. I will address them first.

Fixed the NaN issue. HADDPS has the same problem, which will be fixed in another pr.

File: src/dynarec/rv64/dynarec_rv64_f20f.c [hunk 1]
```diff
@@ -424,6 +424,94 @@ uintptr_t dynarec64_F20F(dynarec_rv64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            s0 = fpu_get_scratch(dyn);
+            s1 = fpu_get_scratch(dyn);
+            // GX->f[0] = GX->f[0] - GX->f[1];
+            FLW(s0, gback, gdoffset + 0);
+            FLW(s1, gback, gdoffset + 4);
+            if (!BOX64ENV(dynarec_fastnan)) {
+                FEQS(x3, s0, s0);
+                FEQS(x4, s1, s1);
+                AND(x5, x3, x4);
+                BEQZ(x5, 4 + 4 * 4);
```
Comment (ptitSeb) [inline src/dynarec/rv64/dynarec_rv64_f20f.c:441]: Are you sure about this offset? you jump 4 opcodes forward? so on iine 448?

File: src/dynarec/rv64/dynarec_rv64_f20f.c [hunk 2]
```diff
@@ -424,6 +424,94 @@ uintptr_t dynarec64_F20F(dynarec_rv64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            }
+            FSW(s0, gback, gdoffset + 0);
+            // GX->f[1] = GX->f[2] - GX->f[3];
+            FLW(s0, gback, gdoffset + 8);
+            FLW(s1, gback, gdoffset + 12);
+            if (!BOX64ENV(dynarec_fastnan)) {
+                FEQS(x3, s0, s0);
+                FEQS(x4, s1, s1);
+                AND(x5, x3, x4);
+                BEQZ(x5, 4 + 4 * 4);
```
Comment (ptitSeb) [inline src/dynarec/rv64/dynarec_rv64_f20f.c:459]: same question about this offset

File: src/dynarec/rv64/dynarec_rv64_f20f.c [hunk 3]
```diff
@@ -424,6 +424,94 @@ uintptr_t dynarec64_F20F(dynarec_rv64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                }
+                FSW(s0, gback, gdoffset + 8);
+                // GX->f[3] = EX->f[2] - EX->f[3];
+                FLW(s0, wback, fixedaddress + 8);
+                FLW(s1, wback, fixedaddress + 12);
+                if (!BOX64ENV(dynarec_fastnan)) {
+                    FEQS(x3, s0, s0);
+                    FEQS(x4, s1, s1);
+                    AND(x5, x3, x4);
+                    BEQZ(x5, 4 + 4 * 4);
```
Comment (ptitSeb) [inline src/dynarec/rv64/dynarec_rv64_f20f.c:502]: same question

File: src/dynarec/rv64/dynarec_rv64_f20f.c [hunk 4]
```diff
@@ -424,6 +424,94 @@ uintptr_t dynarec64_F20F(dynarec_rv64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                FSW(s0, gback, gdoffset + 12);
+            } else {
+                // GX->f[2] = EX->f[0] - EX->f[1];
+                FLW(s0, wback, fixedaddress + 0);
+                FLW(s1, wback, fixedaddress + 4);
+                if (!BOX64ENV(dynarec_fastnan)) {
+                    FEQS(x3, s0, s0);
+                    FEQS(x4, s1, s1);
+                    AND(x5, x3, x4);
+                    BEQZ(x5, 4 + 4 * 4);
```
Comment (ptitSeb) [inline src/dynarec/rv64/dynarec_rv64_f20f.c:484]: same question

File: src/dynarec/rv64/dynarec_rv64_f20f.c (hunk 1 above)
Comment (zengdage) [reply src/dynarec/rv64/dynarec_rv64_f20f.c:441]: Yes. 

1. If x5 is zero, it means that s0 or s1 is NaN, and jump to line 448 to check whether s1 is NaN or not.
2. If s1 isn't NaN, so s0 is NaN, then write s0 to destination. If s1 is NaN, so copy s1 to s0, then write s0 to destination.

Comment (zengdage) [reply src/dynarec/rv64/dynarec_rv64_f20f.c:441]: And I also referenced the VSUBPS case of dynarec_rv64_avx_0f.c

Comment (ptitSeb) [reply src/dynarec/rv64/dynarec_rv64_f20f.c:441]: Ok, LGTM
